go/store/types/value_store.go

```diff
@@ -563,58 +561,107 @@ func makeBatches(hss []hash.HashSet, count int) [][]hash.Hash {
 	return res
 }
 
+type GCMode int
+
+const (
+	GCModeDefault GCMode = iota
+	GCModeFull
+)
+
 // GC traverses the ValueStore from the root and removes unreferenced chunks from the ChunkStore
-func (lvs *ValueStore) GC(ctx context.Context, oldGenRefs, newGenRefs hash.HashSet, safepointF func() error) error {
+func (lvs *ValueStore) GC(ctx context.Context, mode GCMode, oldGenRefs, newGenRefs hash.HashSet, safepointF func() error) error {
 	lvs.versOnce.Do(lvs.expectVersion)
 
 	lvs.transitionToOldGenGC()
 	defer lvs.transitionToNoGC()
 
-	if gcs, ok := lvs.cs.(chunks.GenerationalCS); ok {
+	gcs, gcsOK := lvs.cs.(chunks.GenerationalCS)
+	collector, collectorOK := lvs.cs.(chunks.ChunkStoreGarbageCollector)
+
+	if gcsOK && collectorOK {
 		oldGen := gcs.OldGen()
 		newGen := gcs.NewGen()
 
-		err := newGen.BeginGC(lvs.gcAddChunk)
+		var oldGenHasMany chunks.HasManyF
+		switch mode {
+		case GCModeDefault:
+			oldGenHasMany = oldGen.HasMany
+		case GCModeFull:
+			oldGenHasMany = unfilteredHashFunc
+		default:
+			return fmt.Errorf("unsupported GCMode %v", mode)
+		}
+
+		err := collector.BeginGC(lvs.gcAddChunk)
 		if err != nil {
 			return err
 		}
 
 		root, err := lvs.Root(ctx)
 		if err != nil {
-			newGen.EndGC()
+			collector.EndGC()
 			return err
 		}
 
 		if root == (hash.Hash{}) {
 			// empty root
-			newGen.EndGC()
+			collector.EndGC()
 			return nil
 		}
 
-		oldGenRefs, err = oldGen.HasMany(ctx, oldGenRefs)
+		oldGenRefs, err = oldGenHasMany(ctx, oldGenRefs)
 		if err != nil {
 			return err
 		}
 
 		newGenRefs.Insert(root)
 
-		err = lvs.gc(ctx, oldGenRefs, oldGen.HasMany, newGen, oldGen, nil, func() hash.HashSet {
+		var oldGenFinalizer, newGenFinalizer chunks.GCFinalizer
+		oldGenFinalizer, err = lvs.gc(ctx, oldGenRefs, oldGenHasMany, collector, oldGen, nil, func() hash.HashSet {
 			n := lvs.transitionToNewGenGC()
 			newGenRefs.InsertAll(n)
 			return make(hash.HashSet)
 		})
 		if err != nil {
-			newGen.EndGC()
+			collector.EndGC()
+			return err
+		}
+
+		var newFileHasMany chunks.HasManyF
+		newFileHasMany, err = oldGenFinalizer.AddChunksToStore(ctx)
+		if err != nil {
+			collector.EndGC()
+			return err
+		}
+
+		if mode == GCModeDefault {
+			oldGenHasMany = oldGen.HasMany
+		} else {
+			oldGenHasMany = newFileHasMany
+		}
+
+		newGenFinalizer, err = lvs.gc(ctx, newGenRefs, oldGenHasMany, collector, newGen, safepointF, lvs.transitionToFinalizingGC)
+		if err != nil {
+			collector.EndGC()
 			return err
 		}
 
-		err = lvs.gc(ctx, newGenRefs, oldGen.HasMany, newGen, newGen, safepointF, lvs.transitionToFinalizingGC)
-		newGen.EndGC()
+		err = newGenFinalizer.SwapChunksInStore(ctx)
 		if err != nil {
+			collector.EndGC()
 			return err
 		}
 
-	} else if collector, ok := lvs.cs.(chunks.ChunkStoreGarbageCollector); ok {
+		if mode == GCModeFull {
+			err = oldGenFinalizer.SwapChunksInStore(ctx)
+			if err != nil {
+				collector.EndGC()
+				return err
+			}
+		}
+
+		collector.EndGC()
+	} else if collectorOK {
```

Comment:
A little hard to follow the logic due to the length of the function. Possible to refactor into gcsOK func and a !gcsOK func?

Reply:
I agree it's a bit hard to follow, but I want to find a better way to refactor it rather than split them up, because the two branches are very intimately tied to each other and need to both be visited anytime something changes. I'll try to take another pass, but in a different PR.